modules/ml/src/main/java/org/apache/ignite/ml/trainers/TrainerTransformers.java

```diff
@@ -0,0 +1,369 @@
+            public <K, V> ModelsComposition fit(
+                DatasetBuilder<K, V> datasetBuilder,
+                IgniteBiFunction<K, V, Vector> featureExtractor,
+                IgniteBiFunction<K, V, L> lbExtractor) {
+                datasetBuilder.upstreamTransformersChain().setSeed(
+                    transformationSeed == null
+                        ? new Random().nextLong()
+                        : transformationSeed);
+
+                return runOnEnsemble(
+                    (db, i, fe) -> (() -> trainer.fit(db, fe, lbExtractor)),
+                    datasetBuilder,
+                    ensembleSize,
+                    subsampleRatio,
+                    featureVectorSize,
+                    maxFeaturesCntPerMdl,
+                    featureExtractor,
+                    aggregator,
+                    environment);
+            }
+
+            @Override
+            protected boolean checkState(ModelsComposition mdl) {
+                return mdl.getModels().stream().allMatch(m -> trainer.checkState((M)m));
+            }
+
+            @Override
+            protected <K, V> ModelsComposition updateModel(
+                ModelsComposition mdl,
+                DatasetBuilder<K, V> datasetBuilder,
+                IgniteBiFunction<K, V, Vector> featureExtractor,
+                IgniteBiFunction<K, V, L> lbExtractor) {
+                return runOnEnsemble(
+                    (db, i, fe) -> (() -> trainer.updateModel(
+                        ((ModelWithMapping<Vector, Double, M>)mdl.getModels().get(i)).model(),
+                        db,
+                        fe,
+                        lbExtractor)),
+                    datasetBuilder,
+                    ensembleSize,
+                    subsampleRatio,
+                    featureVectorSize,
+                    maxFeaturesCntPerMdl,
+                    featureExtractor,
+                    aggregator,
+                    environment);
+            }
+        };
+    }
+
+    /**
+     * This method accepts function which for given dataset builder and index of model in ensemble generates
+     * task of training this model.
+     *
+     * @param trainingTaskGenerator Training test generator.
+     * @param datasetBuilder Dataset builder.
+     * @param ensembleSize Size of ensemble.
+     * @param subsampleRatio Ratio (subsample size) / (initial dataset size).
+     * @param aggregator Aggregator of models.
+     * @param environment Environment.
+     * @param <K> Type of keys in dataset builder.
+     * @param <V> Type of values in dataset builder.
+     * @param <M> Type of model.
+     * @return Composition of models trained on bagged dataset.
+     */
+    private static <K, V, M extends Model<Vector, Double>> ModelsComposition runOnEnsemble(
+        IgniteTriFunction<DatasetBuilder<K, V>, Integer, IgniteBiFunction<K, V, Vector>, IgniteSupplier<M>> trainingTaskGenerator,
+        DatasetBuilder<K, V> datasetBuilder,
```

Comment:
It looks like a `DatasetBuilder` is used here just as a container for transformers chain. Are you sure that it's a right approach?

Reply:
I wouldn't say it is just a container for chain: it is also needed for passing into `trainingTaskGenerator`.